libs/sdk-core/src/greenlight/node_api.rs

```diff
@@ -551,6 +585,117 @@ impl NodeAPI for Greenlight {
+        let last_hop = invoice.routing_hints.first().and_then(|rh| rh.hops.first());
+        let mut client: node::ClnClient = self.get_node_client().await?;
+
+        // We first calculate for each channel the max amount to pay (at the receiver)
+        let (mut max_amount_per_channel, payment_path) = self
+            .max_amount_to_send(Some(hex::decode(invoice.payee_pubkey)?), max_hops, last_hop)
+            .await?;
+        info!("send_pay: route: {:?}", payment_path);
+
+        // Calculate the total amount to pay
+        let total: u64 = max_amount_per_channel.iter().map(|m| m.amount_msat).sum();
+
+        // Sort the channels by max amount descending so we can build the route in a way that it
+        // drains the largest channels first
+        max_amount_per_channel.sort_by(|m1, m2| m2.amount_msat.cmp(&m1.amount_msat));
+
+        let amount_to_pay = match invoice.amount_msat {
+            Some(amount) => Ok(amount),
+            None => Err(NodeError::Generic(anyhow!("Invoice has no amount"))),
+        }?;
+
+        if amount_to_pay > total {
+            return Err(NodeError::RouteNotFound(anyhow!(
+                "amount too high, max amount is {}",
+                total
+            )));
+        }
+
+        // This is needed in greenlight for the signer to recognize this invoice.
+        client
+            .pre_approve_invoice(pb::cln::PreapproveinvoiceRequest {
+                bolt11: Some(bolt11.clone()),
+            })
+            .await?;
+
+        // We need to allocate a part id for each part that we are sending.
+        let mut part_id = 1;
+        // The total amount we sent
+        let mut amount_sent = 0;
+        // The total amount received
+        let mut amount_received = 0;
+
+        // The algorithm goes over each channel and drains it until the recived amount
+        // equals to the amount to pay defined in the bolt11 invoice.
+        for max in max_amount_per_channel {
+            // calculating the incoming amount for the remaining amount to pay.
+            let left_to_pay = amount_to_pay - amount_received;
+            // Whether we draining the whole channel balance or only what is left to pay
+            let to_pay = match left_to_pay > max.amount_msat {
+                true => max.amount_msat,
+                false => left_to_pay,
+            };
+
+            // We convert our payment path to an actual route that can be sent to the node.
+            // This requires calculating the right fees and cltv delta in each hop.
+            let (route, sent) = convert_to_send_pay_route(
+                payment_path.clone(),
+                to_pay,
+                invoice.min_final_cltv_expiry_delta,
+            );
+            info!(
+                "send_pay route to pay: {:?}, received_amount = {}",
+                route, to_pay
+            );
+
+            // We send the part using the node API
+            client
+                .send_pay(pb::cln::SendpayRequest {
```

Comment:
We may want to check the outcome of this methodcall has a status that is `pending`. Otherwise we would have already completed the payment. That shouldn't happen, but doesn't hurt either.

Reply:
If there is no practical implications I prefer not to complicate?

Reply:
Yeah makes sense.